File: balanced-binary-tree/balanced-binary-tree.cc
#include <bits/stdc++.h>
using namespace std;

struct TreeNode {
  int val;
  TreeNode *left;
  TreeNode *right;
  TreeNode(int x) : val(x), left(NULL), right(NULL) {}
};
typedef TreeNode *node;
class Solution {
public:
  bool isBalanced(node root) {
    if(!root) return true;
    bool left = isBalanced(root->left), right = isBalanced(root->right);
    int left_dep = root->left ? root->left->val : 0,
        right_dep = root->right ? root->right->val : 0;
    root->val = max(left_dep, right_dep) + 1;
    return left && right && abs(left_dep-right_dep)<2;
  }
};

const int N = 7;
int main(int argc, const char *argv[])
{
  int A[N+1] = 
  { 0,
    1,
    0,2,
    0,0,0,3,
  };
  node nodes[N+1] = { 0 };
  for(int i=1; i<N+1; i++) {
    if(A[i]) nodes[i] = new TreeNode(A[i]);
  }
  for(int i=1; i<N+1; i++) {
    if(2*i < N+1 && nodes[i]) nodes[i]->left = nodes[2*i];
    if(2*i+1 < N+1 && nodes[i]) nodes[i]->right = nodes[2*i+1];
  }
  puts("level order");
  int lv = 0;
  while((1<<lv) <= N) ++lv;
  for(int i=0; i<lv; i++) {
    for(int j=0; j<(1<<i) && (1<<i)+j < N+1; j++) 
      if(nodes[(1<<i)+j]) printf("%d ", nodes[(1<<i)+j]->val);
      else printf("# ");
    puts("");
  }
  puts("--------");

  Solution so;
  cout << so.isBalanced(nodes[1]) << endl;
  return 0;
}

File: rs/src/quests/balanced_binary_tree.rs
/**
 * [110] Balanced Binary Tree
 *
 * Given a binary tree, determine if it is height-balanced.
For this problem, a height-balanced binary tree is defined as:

a binary tree in which the left and right subtrees of every node differ in height by no more than 1.

 
Example 1:
Input: root = [3,9,20,null,null,15,7]
Output: true

Example 2:
Input: root = [1,2,2,3,3,null,null,4,4]
Output: false

Example 3:
Input: root = []
Output: true

 
Constraints:

	The number of nodes in the tree is in the range [0, 5000].
	-104 <= Node.val <= 104

 */
pub struct Solution {}
use super::super::util::tree::{TreeNode, to_tree};

// submission codes start here

// Definition for a binary tree node.
// #[derive(Debug, PartialEq, Eq)]
// pub struct TreeNode {
//   pub val: i32,
//   pub left: Option<Rc<RefCell<TreeNode>>>,
//   pub right: Option<Rc<RefCell<TreeNode>>>,
// }
// 
// impl TreeNode {
//   #[inline]
//   pub fn new(val: i32) -> Self {
//     TreeNode {
//       val,
//       left: None,
//       right: None
//     }
//   }
// }
use std::rc::Rc;
use std::cell::RefCell;
impl Solution {
    pub fn is_balanced(root: Option<Rc<RefCell<TreeNode>>>) -> bool {
        if let Some(root) = root {
            let mut r = root.borrow_mut();
            let left = r.left.take();
            let right = r.right.take();
            if Self::is_balanced(left.clone()) && Self::is_balanced(right.clone()) {
                let hl = left.map(|x| x.borrow().val).unwrap_or(0);
                let hr = right.map(|x| x.borrow().val).unwrap_or(0);
                r.val = 1 + std::cmp::max(hl, hr);
                (hl - hr).abs() <= 1
            } else { false }
        } else { true }
    }
}

// submission codes end

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() {
        assert!(Solution::is_balanced(tree![3,9,20,null,null,15,7]));
        assert!(!Solution::is_balanced(tree![1,2,2,3,3,null,null,4,4]));
    }
}
